File: StaffAdministrationCPP/Employee.cpp
//
// Created by stepa on 11/28/2024.
//

#include "Employee.h"
#include <stddef.h>
#include <iostream>


Employee::Employee(string name, float yearly_salary, department department)
{
    Name = name;
    int words = 0;
    int lenOfSentence = name.size();
    for (int i = 0; i < lenOfSentence; i++)
    {if (name[i] == ' '){words++;}}
    words = words + 1;
    cout << "No. of words = " << words << endl;
    if (name.size() <= 0 || name.empty() || words <= 1)
    {
        throw new std::invalid_argument("Invalid Employee name");
    }

    YearlySalary = yearly_salary;
    Department = department;
}


float Employee::CalculateMonthlySalary()
{
    float monthly_salary = YearlySalary / 12;
    return monthly_salary;
}

void Employee::DisplayInformation()
{
    cout << "Name: " << Name << " Monthly salary: " << CalculateMonthlySalary() << " Department: " << Department <<
        endl;
}

File: StaffAdministrationCsharp/StaffAdministrationCsharp/Employee.cs
namespace StaffAdministrationCsharp;

public class Employee
{
    public string Name;
    public decimal YearlySalary;
    public Department Department;
    public int Words;
    
    
    public Employee(string name, decimal yearlySalary, Department department)
    {
        Name = name;
        if (name == null || name == String.Empty) { throw new ArgumentException("Name cannot be empty"); }
        Words = name.Split(' ').Length;
        if (Words < 2) { throw new ArgumentException("Write Name and Surname"); }
        YearlySalary = yearlySalary;
        Department = department;
        CompanyDepartment(false);
    }


    public decimal CalculateMonthlySalary()
    {
        decimal monthlySalary = YearlySalary / 12;
        return monthlySalary;
    }

    public void DisplayInformation()
    {
        Console.WriteLine($"Name: {Name}, Monthly salary: {CalculateMonthlySalary()}, Department: {Department}");
    }

    public void CompanyDepartment(bool boss)
    {
        if (boss) { Department = Department.Company; }
        else if (!boss)
        {
            if (Department == Department.Company)
            {
                throw new ArgumentException("Only boss can have company department");
            }
        }
    }

}